src/backends/tensorrt/plan_backend.cc

```diff
@@ -760,128 +800,352 @@ PlanBackend::Context::InitializeShapeInputBinding(
-        binding_index, profile_index, nvinfer1::OptProfileSelector::kMAX);
-    context.min_shapes_[io_index] = engine_->getProfileShapeValues(
-        binding_index, profile_index, nvinfer1::OptProfileSelector::kMIN);
-    context.opt_shapes_[io_index] = engine_->getProfileShapeValues(
-        binding_index, profile_index, nvinfer1::OptProfileSelector::kOPT);
-
-    if (!context.context_->setInputShapeBinding(
-            binding_index, context.max_shapes_[io_index])) {
-      return Status(
-          Status::Code::INTERNAL,
-          "trt failed to set the input shape binding for '" + input_name +
-              "' for " + name_ + ".");
+    if (max_byte_size != NO_BATCHING) {
+      // Allocate CUDA memory. Use cudaHostAlloc if zero copy supported. We rely
+      // on buffer_bindings_ being non-nullptr to indicate that the buffer has
+      // been correctly initalized so even for zero-sized tensors always
+      // allocate something.
+      bool zero_copy_support = false;
+      RETURN_IF_ERROR(
+          SupportsIntegratedZeroCopy(gpu_device_, &zero_copy_support));
+      void* buffer = nullptr;
+      cudaError_t err = cudaSuccess;
+      if (zero_copy_support) {
+        err = cudaHostAlloc(
+            &buffer, std::max((int64_t)1, max_byte_size), cudaHostAllocMapped);
+      } else {
+        err = cudaMalloc(&buffer, std::max((int64_t)1, max_byte_size));
+      }
+      if (err != cudaSuccess) {
+        return Status(
+            Status::Code::INTERNAL, "unable to allocate memory for input '" +
+                                        input_name + " for " + name_ + ": " +
+                                        cudaGetErrorString(err));
+      }
+
+      io_binding_info.byte_size_ = max_byte_size;
+      io_binding_info.buffer_ = buffer;
+      io_binding_info.device_buffer_ = buffer;
+      if (zero_copy_support) {
+        io_binding_info.memory_type_ = TRITONSERVER_MEMORY_CPU_PINNED;
+        io_binding_info.memory_type_id_ = 0;
+        err = cudaHostGetDevicePointer(
+            &io_binding_info.device_buffer_, io_binding_info.buffer_, 0);
+        if (err != cudaSuccess) {
+          return Status(
+              Status::Code::INTERNAL,
+              "unable to get mapped device address for input '" + input_name +
+                  " for " + name_ + ": " + cudaGetErrorString(err));
+        }
+      } else {
+        io_binding_info.memory_type_ = TRITONSERVER_MEMORY_GPU;
+        io_binding_info.memory_type_id_ = gpu_device_;
+      }
+
+      // Set buffer bindings of all optimization profile since buffer is
+      // allocated
+      for (auto& trt_context : trt_contexts_) {
+        auto binding_index =
+            num_expected_bindings_ * trt_context.first + io_index;
+        buffer_bindings_[s][binding_index] = io_binding_info.device_buffer_;
+      }
     }
+  }
+  return Status::Success;
+}
+
+Status
+PlanBackend::Context::InitializeExecuteInputBinding(
```

Comment:
The Github diff is acting weird so it is hard to compare your change. Maybe instead of iterating the `io_binding_infos_` in the `InitializeXXX()` functions, would that be better if you move it to line 620 so that it becomes:
```
for (int s = 0; s < num_copy_streams_; s++) {
   next_buffer_binding_set_ = s;
  RETURN_IF_ERROR(
      context->InitializeConfigShapeInputBindings(Config().input()));
  RETURN_IF_ERROR(
      context->InitializeConfigExecuteInputBindings(Config().input()));
   ...
}
```
You may utilize `next_buffer_binding_set_` so the `InitializeXXX()` function can set the proper buffer set without changing the signature.